### PR TITLE
Market Resolution API

### LINKED ISSUE #559: Feature request: Add an API route to resolve markets
This was discussed in the Discord, adding an issue here just for tracking

### PR DESCRIPTION
**Pull request by bcongdon**
Resolves #559

Notes:
- I added additional validation to `resolve-market` to prevent some invalid states, like resolving to a FR outcome that doesn't exist, or resolving `MULTI` with percentages that don't add up to 100%.
- Currently, we're using `contractId` for market resolution. Eventually, we should move to `slug`.

### CODE REVIEW

File: functions/src/resolve-market.ts [hunk 1]
```diff
@@ -74,6 +78,20 @@ export const resolvemarket = newEndpoint(opts, async (req, auth) => {
   const creator = await getUser(creatorId)
   if (!creator) throw new APIError(500, 'Creator not found')
 
+  if (contract.outcomeType === 'FREE_RESPONSE') {
+    if (resolutions) {
+      Object.keys(resolutions).forEach((outcome) =>
+        validateFreeResponseOutcome(contract, outcome)
+      )
+      const pctSum = sum(Object.values(resolutions))
+      if (pctSum !== 100) {
```
Comment (mqp) [inline functions/src/resolve-market.ts:87]: This is going to be incorrect due to floating point error.

File: functions/src/resolve-market.ts [hunk 2]
```diff
@@ -74,6 +78,20 @@ export const resolvemarket = newEndpoint(opts, async (req, auth) => {
...
+        throw new APIError(400, 'Resolution percentages must sum to 100')
+      }
+    } else if (!isNaN(+outcome)) {
```
Comment (mqp) [inline functions/src/resolve-market.ts:90]: `z.number` already excludes `NaN`.

File: web/lib/api/proxy.ts [hunk 3]
```diff
@@ -41,7 +41,12 @@ export const fetchBackend = (req: NextApiRequest, name: string) => {
     'Origin',
   ])
   const hasBody = req.method != 'HEAD' && req.method != 'GET'
-  const opts = { headers, method: req.method, body: hasBody ? req : undefined }
+  const body = req.body ? JSON.stringify(req.body) : req
```
[the remark this answers is not in the document]
Comment (mqp) [reply web/lib/api/proxy.ts:44]: I see, I hadn't thought about this.

**Review by mqp — COMMENTED** [review]
Seems highly reasonable, I will test, make any necessary changes, and deploy.

File: functions/src/resolve-market.ts [hunk 4]
```diff
@@ -74,6 +78,20 @@ export const resolvemarket = newEndpoint(opts, async (req, auth) => {
   const creator = await getUser(creatorId)
   if (!creator) throw new APIError(500, 'Creator not found')
 
+  if (contract.outcomeType === 'FREE_RESPONSE') {
```
Comment (mqp) [inline functions/src/resolve-market.ts:81]: Wouldn't it make more sense for this whole block to be in `getResolutionParams`?

File: functions/src/resolve-market.ts (hunk 1 above)
Comment (bcongdon) [reply functions/src/resolve-market.ts]: Fixed, using a tolerance of 0.1 now.

File: functions/src/resolve-market.ts (hunk 2 above)
Comment (bcongdon) [reply functions/src/resolve-market.ts]: This was intended to check "is outcome a number". Switched to using lodash's `isNumber`.

**Comment by mqp** [comment]
I just went ahead and refactored it so that I liked it better. Basically, try to take advantage of Typescript more -- if you correctly use type narrowing then you shouldn't have to be doing stuff like checking `'resolutions` in freeResponseParams` all over the place.

Anyway this seems to work great so I am merging and deploying.

**Pull request merged**